### PR TITLE
Add file check & reuse protobuf conversion functions

### PR DESCRIPTION
**Pull request by awjuliani**
* Add check for file existence in demonstration loader. 
* Reuse protobuf reading functions between environment and demonstration loader.

### CODE REVIEW

File: ml-agents/mlagents/envs/utilities.py [hunk 1]
```diff
@@ -1,19 +1,94 @@
...
     and optionally converts it to grey scale
+    :param grey_scale: Whether to convert the image to greyscale.
     :param image_bytes: input byte array corresponding to image
     :return: processed numpy array of observation from environment
     """
     s = bytearray(image_bytes)
     image = Image.open(io.BytesIO(s))
     s = np.array(image) / 255.0
-    if gray_scale:
+    if grey_scale:
```
Comment (vincentpierre) [inline ml-agents/mlagents/envs/utilities.py]: Is this renaming necessary ? It is gray scale everywhere else

File: ml-agents/mlagents/envs/utilities.py [hunk 2]
```diff
@@ -1,19 +1,94 @@
...
         s = np.mean(s, axis=2)
         s = np.reshape(s, [s.shape[0], s.shape[1], 1])
     return s
+
+
+def brain_param_proto_to_obj(brain_param_proto):
```
Comment (vincentpierre) [inline ml-agents/mlagents/envs/utilities.py]: This should be a constructor or a factory in BrainParameters

File: ml-agents/mlagents/envs/utilities.py [hunk 3]
```diff
@@ -1,19 +1,94 @@
...
+        "numStackedVectorObservations": brain_param_proto.num_stacked_vector_observations,
+        "cameraResolutions": resolution,
+        "vectorActionSize": brain_param_proto.vector_action_size,
+        "vectorActionDescriptions": brain_param_proto.vector_action_descriptions,
+        "vectorActionSpaceType": brain_param_proto.vector_action_space_type
+    })
+    return brain_params
+
+
+def agent_info_proto_to_brain_info(agent_info_list, brain_params):
```
Comment (vincentpierre) [inline ml-agents/mlagents/envs/utilities.py]: This should be a constructor or a factory on BrainInfo.

File: ml-agents/mlagents/envs/utilities.py [hunk 4]
```diff
@@ -1,19 +1,94 @@
...
+    """
+    Converts list of agent infos to BrainInfo.
```
Comment (vincentpierre) [inline ml-agents/mlagents/envs/utilities.py]: Can we make the name of the method more appropriate ?

File: ml-agents/mlagents/envs/utilities.py [hunk 5]
```diff
@@ -1,19 +1,94 @@
...
+    if grey_scale:
         s = np.mean(s, axis=2)
         s = np.reshape(s, [s.shape[0], s.shape[1], 1])
     return s
+
+
+def brain_param_proto_to_obj(brain_param_proto):
+    """
+    Converts brain parameter proto to BrainParameter object.
+    :param brain_param_proto: protobuf object.
```
Comment (vincentpierre) [inline ml-agents/mlagents/envs/utilities.py]: I think the name of the method could be better

File: ml-agents/mlagents/envs/utilities.py [hunk 6]
```diff
@@ -1,19 +1,94 @@
...
 import io
+import logging
 
+from mlagents.envs.brain import BrainParameters, BrainInfo
 
-def process_pixels(image_bytes, gray_scale):
+logger = logging.getLogger("mlagents.envs")
+
+
+def process_pixels(image_bytes, grey_scale):
```
Comment (vincentpierre) [inline ml-agents/mlagents/envs/utilities.py]: If `process_pixels` is only used in `brain_param_proto_to_obj` and we move `brain_param_proto_to_obj` inside of BrainInfo maybe it should not be in `utilities.py` anymore?

File: ml-agents/mlagents/envs/environment.py [hunk 7]
```diff
@@ -507,7 +466,8 @@ def _generate_reset_input(self, training, config) -> UnityRLInput:
         rl_in.command = 1
         return self.wrap_unity_input(rl_in)
 
-    def send_academy_parameters(self, init_parameters: UnityRLInitializationInput) -> UnityRLInitializationOutput:
+    def send_academy_parameters(self,
```
Comment (vincentpierre) [inline ml-agents/mlagents/envs/environment.py]: Shouldn't the indentation be 
```
def send_academy_parameters(
         self, init_parameters: UnityRLInitializationInput) -> UnityRLInitializationOutput:
```
?

File: ml-agents/mlagents/envs/environment.py [hunk 8]
```diff
@@ -434,47 +430,10 @@ def _get_state(self, output: UnityRLOutput) -> (AllBrainInfo, bool):
...
-                agents=[x.id for x in agent_info_list],
-                local_done=[x.done for x in agent_info_list],
-                vector_action=np.array([x.stored_vector_actions for x in agent_info_list]),
-                text_action=[x.stored_text_actions for x in agent_info_list],
-                max_reached=[x.max_step_reached for x in agent_info_list],
-                action_mask=mask_actions
-                )
+        for brain_name in output.agentInfos:
+            agent_info_list = output.agentInfos[brain_name].value
+            _data[brain_name] = agent_info_proto_to_brain_info(agent_info_list,
```
Comment (vincentpierre) [inline ml-agents/mlagents/envs/environment.py]: I think the indentation is off : 
```
 agent_info_proto_to_brain_info(
           agent_info_list, self.brains[brain_name])
```

File: ml-agents/mlagents/envs/environment.py [hunk 9]
```diff
@@ -361,17 +353,19 @@ def step(self,  vector_action=None, memory=None, text_action=None, value=None) -
                         "The brain {0} expected {1} text_action but was given {2}".format(
                             b, n_agent, len(text_action[b])))
                 if not ((self._brains[b].vector_action_space_type == "discrete" and len(
-                        vector_action[b]) == n_agent * len(self._brains[b].vector_action_space_size)) or
-                            (self._brains[b].vector_action_space_type == "continuous" and len(
-                                vector_action[b]) == self._brains[b].vector_action_space_size[0] * n_agent)):
+                        vector_action[b]) == n_agent * len(
```
Comment (mmattar) [inline ml-agents/mlagents/envs/environment.py]: this block condition is super confusing ... can we use intermediate variables to separate the different checks.

For example:
```
total_space_size = n_agent * len(self._brains[b].vector_action_space_size)
discrete_space_check = (self._brains[b].vector_action_space_type == "discrete" and len(vector_action[b]) == total_space_size
if not (discrete_space_check or ...
```

Comment (mmattar) [reply ml-agents/mlagents/envs/environment.py]: These temp vars can now be reused in the error messages that follow.

Comment (mmattar) [reply ml-agents/mlagents/envs/environment.py]: Re-write / re-organize block till line 368 for readability.

File: ml-agents/mlagents/envs/environment.py (hunk 7 above)
Comment (awjuliani) [reply ml-agents/mlagents/envs/environment.py]: Pycharm doesn't like it that way.

File: ml-agents/mlagents/envs/environment.py (hunk 8 above)
Comment (awjuliani) [reply ml-agents/mlagents/envs/environment.py]: Pycharm doesn't like it that way.

**Comment by awjuliani** [comment]
@mmattar @vincentpierre Comments addressed.

File: ml-agents/mlagents/envs/brain.py [hunk 10]
```diff
@@ -20,25 +28,92 @@ def __init__(self, visual_observation, vector_observation, text_observations, me
...
+
+    @staticmethod
+    def from_agent_proto(agent_info_list, brain_params):
+        """
+        Converts list of agent infos to BrainInfo.
+        """
+        vis_obs = []
+        for i in range(brain_params.number_visual_observations):
+            obs = [BrainInfo.process_pixels(x.visual_observations[i],
+                                  brain_params.camera_resolutions[i]['blackAndWhite'])
```
Comment (mmattar) [inline ml-agents/mlagents/envs/brain.py]: indentation for lines 56/57 seems off.

File: ml-agents/mlagents/trainers/demo_loader.py [hunk 11]
```diff
@@ -117,6 +54,8 @@ def load_demonstration(file_path):
     """
     INITIAL_POS = 33
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/demo_loader.py]: Can we have a comment

**Pull request merged**